File: Number of Ways Where Square of Number Is Equal to Product of Two Numbers/Number of Ways Where Square of Number Is Equal to Product of Two Numbers.cpp
class Solution {
public:
    int numTriplets(vector<int>& nums1, vector<int>& nums2) {
        return count(nums1,nums2)+count(nums2,nums1);
    }
    int count(vector<int>& nums1,vector<int>& nums2){
        int res=0;
        unordered_map<int,int> m;
        for(int n:nums2)++m[n];
        for(int a:nums1){
            long target=(long)a*a;
            for(auto& [k,v]:m){
                if(target%k!=0||!m.count(target/k))continue;
                if(target/k==k)res+=v*(v-1);
                else
                    res+=v*m[target/k];
            }
        }
        return res/2;
    }
};

File: Solutions/Number of Ways Where Square of Number Is Equal to Product of Two Numbers/Number of Ways Where Square of Number Is Equal to Product of Two Numbers.java
class Solution {
    public int numTriplets(int[] nums1, int[] nums2) {
        return (int)(count(nums1,nums2)+count(nums2,nums1));
    }
    private long count(int[] nums1,int[] nums2){
        long res=0;
        HashMap<Long,Long> map=new HashMap<>();
        for(long n2:nums2){
            long tmp=map.getOrDefault(n2,0l);
            ++tmp;
            map.put(n2,tmp);
        }
        for(long n1:nums1){
            long target=n1*n1;
            for(var entry:map.entrySet()){
                long key=entry.getKey();
                long value=entry.getValue();
                if(target%key!=0||!map.containsKey(target/key))
                    continue;
                if(target/key==key)
                    res+=value*(value-1);
                else
                    res+=value*map.get(target/key);
            }
        }
        return res/2;
    }
}